### PR TITLE
Fixed wasm event listeners by using js callbacks.

### PR DESCRIPTION
**Pull request by norunners**

### CODE REVIEW

File: dom_wasm.go [hunk 1]
```diff
@@ -1088,14 +1088,16 @@ func (w *window) Stop() {
 
 // TODO reuse util.EventTarget
 
-func (w *window) AddEventListener(typ string, useCapture bool, listener func(Event)) func(o js.Value) {
-	wrapper := func(o js.Value) { listener(wrapEvent(o)) }
-	w.Call("addEventListener", typ, wrapper, useCapture)
-	return wrapper
+func (w *window) AddEventListener(typ string, useCapture bool, listener func(Event)) js.Callback {
+	wrapper := func(o []js.Value) { listener(wrapEvent(o[0])) }
+	cb := js.NewCallback(wrapper)
```
Comment (bketelsen) [inline dom_wasm.go:1093]: I think these should be NewEventCallback() and accept the callback flags too.

File: dom_wasm.go [hunk 2]
```diff
@@ -1088,14 +1088,16 @@ func (w *window) Stop() {
...
+	w.Call("addEventListener", typ, cb, useCapture)
+	return cb
 }
 
-func (w *window) RemoveEventListener(typ string, useCapture bool, listener func(js.Value)) {
+func (w *window) RemoveEventListener(typ string, useCapture bool, listener js.Callback) {
```
Comment (bketelsen) [inline dom_wasm.go:1098]: there's a comment in the syscall/js code that says removing can only work if you remove the exact object you add, so this should be RemoveEventListener(cb js.Callback)

Comment (bketelsen) [reply dom_wasm.go:1098]: then fix the line below to use the remove func from syscall/js

File: dom_wasm.go [hunk 3]
```diff
@@ -1272,14 +1274,16 @@ func (n *BasicNode) Underlying() js.Value {
 	return n.Object
 }
 
-func (n *BasicNode) AddEventListener(typ string, useCapture bool, listener func(Event)) func(js.Value) {
-	wrapper := func(o js.Value) { listener(wrapEvent(o)) }
-	n.Object.Call("addEventListener", typ, wrapper, useCapture)
-	return wrapper
+func (n *BasicNode) AddEventListener(typ string, useCapture bool, listener func(Event)) js.Callback {
```
Comment (bketelsen) [inline dom_wasm.go:1277]: same here

File: dom_wasm.go [hunk 4]
```diff
@@ -1272,14 +1274,16 @@ func (n *BasicNode) Underlying() js.Value {
...
+	wrapper := func(o []js.Value) { listener(wrapEvent(o[0])) }
+	cb := js.NewCallback(wrapper)
+	n.Object.Call("addEventListener", typ, cb, useCapture)
+	return cb
 }
 
-func (n *BasicNode) RemoveEventListener(typ string, useCapture bool, listener func(js.Value)) {
+func (n *BasicNode) RemoveEventListener(typ string, useCapture bool, listener js.Callback) {
```
Comment (bketelsen) [inline dom_wasm.go:1284]: same here

File: events_wasm.go [hunk 5]
```diff
@@ -409,7 +409,7 @@ type EventTarget interface {
 	// AddEventListener adds a new event listener and returns the
 	// wrapper function it generated. If using RemoveEventListener,
 	// that wrapper has to be used.
-	AddEventListener(typ string, useCapture bool, listener func(Event)) func(js.Value)
-	RemoveEventListener(typ string, useCapture bool, listener func(js.Value))
+	AddEventListener(typ string, useCapture bool, listener func(Event)) js.Callback
```
Comment (bketelsen) [inline events_wasm.go:412]: then fix these interface declarations to match

**Review by bketelsen — CHANGES_REQUESTED** [review]
Great change!  Just a few changes suggested.

File: dom_wasm.go (hunk 2 above)
Comment (norunners) [reply dom_wasm.go:1098]: Please link the comment around callback removal.

Can you provide an example of the change you are expecting?

File: dom_wasm.go (hunk 1 above)
Comment (norunners) [reply dom_wasm.go:1093]: Makes sense, will add `js.EventCallbackFlag` as an argument.

Comment (norunners) [reply dom_wasm.go:1093]: Just curious, what value does adding the `js.EventCallbackFlag` when the dom.Event object already has those methods available?

For example: `event.PreventDefault()`

This seems to introduce two ways to do the same thing.

**Pull request merged**